solr/core/src/java/org/apache/solr/cluster/placement/Cluster.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * <p>A representation of the (initial) cluster state, providing information on which nodes are part of the cluster and a way
+ * to get to more detailed info.
+ *
+ * <p>This instance can also be used as a {@link PropertyValueSource} if {@link PropertyKey}'s need to be specified with
+ * a global cluster target.
+ */
+public interface Cluster extends PropertyValueSource {
+  /**
+   * @return current set of live nodes. Never <code>null</code>, never empty (Solr wouldn't call the plugin if empty
+   * since no useful could then be done).
+   */
+  Set<Node> getLiveNodes();
+
+  /**
+   * <p>Returns info about the given collection if one exists. Because it is not expected for plugins to request info about
+   * a large number of collections, requests can only be made one by one.
+   *
+   * <p>This is also the reason we do not return a {@link java.util.Map} or {@link Set} of {@link SolrCollection}'s here: it would be
+   * wasteful to fetch all data and fill such a map when plugin code likely needs info about at most one or two collections.
+   */
+  Optional<SolrCollection> getCollection(String collectionName) throws IOException;
```

Comment:
Ok. I don't really see a use case for this (interested if you have something specific in mind) but will add. My thinking was that plugins will be interested in the Collection they need to compute placement for and other specific collections if "their" collection properties reference another collection (for example something along the lines of `withCollection`, even though we remove it, a plugin could reimplement).

Reply:
Right, I thought of `withCollection`. But there are other cases where we may want to avoid placing replicas on nodes with replicas from other collections. Eg. at Lucidworks we heavily use several predefined system collections for logging and analytics jobs, and we want to avoid putting search collections on those nodes if possible. This is currently achieved by totally separating the nodes using exclusive policy properties, but a different softer mechanism could be better on resource-constrained clusters - but then you need to know the placements of other collections...

Reply:
I will add this.
To your use case: if the plugin knows the names of the predefined system collections not to colocate new replicas with, it can request them one by one by name. Where an iteration would be useful is if the plugin doesn't want to place replicas on any `Node` with replicas from any collection named `system-*-metric` for example. It would then have to iterate...
I'm always thinking eventually SolrCloud will be scaling to hundreds of thousands collections per cluster so anything that's linear in execution with that number is not acceptable in such a case for placement computation. But I do understand it can be useful in low scale uses.

Reply:
You can use system property rules for ensuring replicas for one collection don't get placed on nodes hosting another collection vs. the new framework trying to compute this by looking at all the nodes hosting a specific collection (which can change anyway). Here's an example using the current framework: 
https://github.com/lucidworks/fusion-cloud-native/blob/master/policy.json#L19

This approach is also simple to reason about and is in-use in a number of production environments. Any ops person will be comfortable with setting JVM props or ENV vars to indicate unique Solr node properties that they want to use for placement decisions.

Reply:
The new framework does not mandate any specific way to do this. The scenario you describe (partitioning nodes by system properties) is totally possible to implement, then one would be using `[PropertyValue.Sysprop](https://github.com/apache/lucene-solr/pull/1684/files#diff-d4ace968d26586330e5b20a0b698e9a7R125)` to implement that.